kong/plugins/grpc-gateway/deco.lua

```diff
@@ -266,7 +267,17 @@ function deco:upstream(body)
       end
     end
   end
-  body = grpc_frame(0x0, pb.encode(self.endpoint.input_type, payload))
+
+  local pok, msg = pcall(pb.encode, self.endpoint.input_type, payload)
+  if not pok or not msg then
+    if msg then
+      ngx.log(ngx.ERR, msg)
```

Comment:
it will be better if we return this error message to the function caller.

Reply:
@attenuation Do you have any data that proves localizing these variables would be beneficial?

Reply:
In Luajit, I try to dump some asm code
```
╭─ouyang@attenuation-home ~/tmp/lua
╰─$ luajit -jdump -e "local x=0; for i=1,1e6 do x=x+i end; print(x)"

---- TRACE 1 start (command line):1
0006  ADDVV    0   0   4
0007  FORL     1 => 0006
---- TRACE 1 IR
0001    int SLOAD  #3    CI
0002 >  num SLOAD  #2    T
0003    num CONV   0001  num.int
0004  + num ADD    0003  0002
0005  + int ADD    0001  +1
0006 >  int LE     0005  +1000000
0007 ------ LOOP ------------
0008    num CONV   0005  num.int
0009  + num ADD    0008  0004
0010  + int ADD    0005  +1
0011 >  int LE     0010  +1000000
0012    int PHI    0005  0010
0013    num PHI    0004  0009
---- TRACE 1 mcode 82
7efc8828ffa9  mov dword [r14-0xec8], 0x1
7efc8828ffb4  cvttsd2si ebp, [rdx+0x8]
7efc8828ffb9  cmp dword [rdx+0x4], 0xfff90000
7efc8828ffc0  jnb 0x7efc8828004c	->0
7efc8828ffc6  xorps xmm7, xmm7
7efc8828ffc9  cvtsi2sd xmm7, ebp
7efc8828ffcd  addsd xmm7, [rdx]
7efc8828ffd1  add ebp, +0x01
7efc8828ffd4  cmp ebp, 0x000f4240
7efc8828ffda  jg 0x7efc88280050	->1
->LOOP:
7efc8828ffe0  xorps xmm6, xmm6
7efc8828ffe3  cvtsi2sd xmm6, ebp
7efc8828ffe7  addsd xmm7, xmm6
7efc8828ffeb  add ebp, +0x01
7efc8828ffee  cmp ebp, 0x000f4240
7efc8828fff4  jle 0x7efc8828ffe0	->LOOP
7efc8828fff6  jmp 0x7efc88280058	->3
---- TRACE 1 stop -> loop

500000500000
╭─ouyang@attenuation-home ~/tmp/lua
╰─$ luajit -jdump -e "x=0; for i=1,1e6 do x=x+i end; print(x)"

---- TRACE 1 start (command line):1
0007  GGET     4   0      ; "x"
0008  ADDVV    4   4   3
0009  GSET     4   0      ; "x"
0010  FORL     0 => 0007
---- TRACE 1 IR
0001    int SLOAD  #2    CI
0002    fun SLOAD  #0    R
0003    tab FLOAD  0002  func.env
0004    int FLOAD  0003  tab.hmask
0005 >  int EQ     0004  +63
0006    p64 FLOAD  0003  tab.node
0007 >  p64 HREFK  0006  "x"  @46
0008 >  num HLOAD  0007
0009    num CONV   0001  num.int
0010  + num ADD    0009  0008
0011    num HSTORE 0007  0010
0012  + int ADD    0001  +1
0013 >  int LE     0012  +1000000
0014 ------ LOOP ------------
0015    num CONV   0012  num.int
0016  + num ADD    0015  0010
0017    num HSTORE 0007  0016
0018  + int ADD    0012  +1
0019 >  int LE     0018  +1000000
0020    int PHI    0012  0018
0021    num PHI    0010  0016
---- TRACE 1 mcode 148
7f883227ff6b  mov dword [r14-0xec8], 0x1
7f883227ff76  cvttsd2si ebp, [rdx]
7f883227ff7a  rorx rbx, [rdx-0x10], 0x2f
7f883227ff81  shr rbx, 0x11
7f883227ff85  mov rdx, [rbx+0x10]
7f883227ff89  cmp dword [rdx+0x34], +0x3f
7f883227ff8d  jnz 0x7f883227004c	->0
7f883227ff93  mov rcx, [rdx+0x28]
7f883227ff97  mov rdi, 0xfffdff8865e76bd8
7f883227ffa1  cmp rdi, [rcx+0x458]
7f883227ffa8  jnz 0x7f883227004c	->0
7f883227ffae  lea rax, [rcx+0x450]
7f883227ffb5  cmp dword [rax+0x4], 0xfff90000
7f883227ffbc  jnb 0x7f883227004c	->0
7f883227ffc2  xorps xmm7, xmm7
7f883227ffc5  cvtsi2sd xmm7, ebp
7f883227ffc9  addsd xmm7, [rax]
7f883227ffcd  movsd [rax], xmm7
7f883227ffd1  add ebp, +0x01
7f883227ffd4  cmp ebp, 0x000f4240
7f883227ffda  jg 0x7f8832270050	->1
->LOOP:
7f883227ffe0  xorps xmm6, xmm6
7f883227ffe3  cvtsi2sd xmm6, ebp
7f883227ffe7  addsd xmm7, xmm6
7f883227ffeb  movsd [rax], xmm7
7f883227ffef  add ebp, +0x01
7f883227fff2  cmp ebp, 0x000f4240
7f883227fff8  jle 0x7f883227ffe0	->LOOP
7f883227fffa  jmp 0x7f8832270058	->3
---- TRACE 1 stop -> loop

500000500000


we can see non local `x` version has more ASM code in Luajit, 
```

Reply:
The microbenchmark you're using does not reflect how such source-level changes influence the actual runtime performance of a bigger system, in which instruction reordering and pipelining are present and can completely negate effects of such local optimizations.  For that reason, I would advise against applying such optimizations unless they have proven to show a noticeable effect on the runtime of the code in question.

I would particularly be against optimizing based on the Lua interpreter optimization tips because our runtime is not the Lua interpreter, and interpreter optimiziation is a thing completely different from compiler optimization.

Let us not add code that is not needed.

Reply:
Localizing variables is a coding style in the OpenResty that can only marginally improve performance. 
For more information, please refer to:
1. https://groups.google.com/g/openresty-en/c/9uM-nvravX4/m/mwca_7M9DAAJ
2. https://github.com/openresty/lua-nginx-module#lua-variable-scope

FYI: I would prefer to follow this coding style as well. However, first we need to come to an agreement.

Reply:
I strongly agree with Hans' opinion.

Readability and maintainability are in most cases just as important, and optimizing code for every last ounce of performance can severely impact those qualities. Generally speaking, optimize only once you are sure that there is or will be a performance bottleneck.

Reply:
I want to emphasize the importance of following established coding style of the project, and I believe small things like these do adds up.

@ms2008 I am not following your comment, what "Readability and maintainability" issues does localization brings specifically?

Reply:
@dndx For example, Hans mentioned [here](https://github.com/Kong/kong-ee/pull/4209#discussion_r1164219047) about the `split` issue, whether it's from Penlight's `stringx.split` or `ngx.re.split`, and we need to jump to the localizing place to confirm. It's not a big deal, but it's annoying.

If we want to keep localizing stuff, we should emphasize that this is just to **follow the coding style of the project not overemphasis on performance related**.
    
Because we don't have any evidence that localizations perform better than globals in Luajit's world.

Reply:
I agree, as a matter of fact I already discussed with @hanshuebner offline and we agree consistent localization is better than arbitrary localization we are doing right now.

@pluveto is this something Lualint can help with?